joey/layers_v2.py

```diff
@@ -391,3 +393,357 @@ def __init__(self, kernel_size, input_size,
+                              space_order=0, dtype=np.float64)
+
+        # Result for pool layer
+        result_dimensions = [SpaceDimension("Result_"+x) for x in dimensions]
+
+        result_func = Function(name="Result_F", shape=result_shape,
+                               dimensions=result_dimensions, space_order=0,
+                               dtype=np.float64)
+
+        output_grad_dimensions = [SpaceDimension(
+            "output_grad"+x) for x in dimensions]
+
+        output_grad = Function(name="outgrad_%s" % name_allocator_func(
+        ), shape=result_shape, dimensions=output_grad_dimensions,
+            space_order=0, dtype=np.float64)
+
+        self._set_padding_result_values(
+            input_func, result_func, value=np.finfo(np.float32).min)
+
+        return (None, input_func, result_func, None, None, output_grad, None)
+
+    @property
+    def stride(self):
+        """Stride of the layer."""
+        return self._stride
+
+    @property
+    def kernel_size(self):
+        """The kernel size of the layer."""
+        return self._kernel_size
+
+    def execute(self, input_data):
+        '''execute implementation'''
+        indices = [slice(0, self._I.shape[0], 1),
+                   slice(0, self._I.shape[1], 1)]
+        [indices.append(slice(self._padding[i],
+                              self._I.data.shape[2+i]-self._padding[i], 1))
+            for i in range(self._dims)]
+
+        self._I.data[tuple(indices)] = input_data
+
+        return super().execute()
+
+    @abstractmethod
+    def equations(self):
+        pass
+
+    @abstractmethod
+    def backprop_equations(self, prev_layer, next_layer):
+        pass
+
+
+class MaxPoolingV2(Pooling):
+    """
+    A Layer/Pooling subclass corresponding to a max pooling layer
+
+    Parameters
+    ----------
+    See Pooling.__doc__.
+    """
+
+    def __init__(self, kernel_size, input_size, dims,
+                 stride, padding, activation, generate_code,
+                 strict_stride_check):
+        self._indices = None
+        self._forward_tmp_constants = None
+        self._backward_tmp_constants = None
+        super().__init__(kernel_size, input_size, dims, alloc, dim_alloc,
+                         stride, padding, activation, generate_code,
+                         strict_stride_check)
+
+    def equations(self):
+        result_dimensions = self._R.dimensions
+
+        input_dims = [result_dimensions[0], result_dimensions[1]]
+        args = []
+        for i in range(0, self._dims):
+            new_dim = SpaceDimension(
+                name='d_kernel'+self._dim_dict.get(self._dims-i, self._dims-i))
+            input_dims.append(
+                result_dimensions[-self._dims + i]*self._stride[i]+new_dim)
+            args.append((new_dim.name + '_M', self._kernel_size[i] - 1))
+
+        eqs = [Eq(self._R[result_dimensions], Max(
+            self._R[result_dimensions], self._I[input_dims], evaluate=False))]
```

Comment:
Could we do this in a way that the equations are actually written down in the superclass (Pooling), and this class only provides the Max function?

Reply:
Let's add this as an issue and leave the refactoring for another PR.

Reply:
Issue #18